File: packages/java/src/main/java/io/opensms/Countries.java
package io.opensms;

import io.opensms.models.Carrier;
import io.opensms.models.Country;
import io.opensms.models.CountryRules;
import io.opensms.models.Route;

import java.util.List;

/** The public country catalog. Accessed as {@code client.countries()}. */
public final class Countries {

    private final ApiTransport t;

    Countries(ApiTransport t) {
        this.t = t;
    }

    /** @return every supported country. */
    public List<Country> list() {
        return t.get("/v1/countries", null, ApiTransport.listOf(Country.class));
    }

    /**
     * @param iso2 ISO2 country.
     * @return carriers in that country.
     */
    public List<Carrier> carriers(String iso2) {
        return t.get("/v1/countries/" + Wire.id(iso2, "iso2") + "/carriers", null, ApiTransport.listOf(Carrier.class));
    }

    /**
     * @param iso2 ISO2 country.
     * @return routable providers in that country.
     */
    public List<Route> routes(String iso2) {
        return t.get("/v1/countries/" + Wire.id(iso2, "iso2") + "/routes", null, ApiTransport.listOf(Route.class));
    }

    /**
     * @param iso2 ISO2 country.
     * @return that country's compliance rules.
     */
    public CountryRules compliance(String iso2) {
        return t.get("/v1/countries/" + Wire.id(iso2, "iso2") + "/compliance", null, ApiTransport.type(CountryRules.class));
    }
}

File: packages/go/countries.go
package opensms

import (
	"context"
	"net/http"
)

// Countries reads the public country catalog, reached as client.Countries.
type Countries struct {
	http *transport
}

// List returns the active countries (GET /v1/countries).
func (r *Countries) List(ctx context.Context) ([]Country, error) {
	return callSlice[Country](ctx, r.http, request{method: http.MethodGet, path: "/v1/countries"})
}

// Carriers returns a country's carriers (GET /v1/countries/{iso2}/carriers).
func (r *Countries) Carriers(ctx context.Context, iso2 string) ([]Carrier, error) {
	if err := requireID("iso2", iso2); err != nil {
		return nil, err
	}
	return callSlice[Carrier](ctx, r.http, request{method: http.MethodGet, path: "/v1/countries/" + escape(iso2) + "/carriers"})
}

// Routes returns a country's routable routes (GET /v1/countries/{iso2}/routes).
func (r *Countries) Routes(ctx context.Context, iso2 string) ([]Route, error) {
	if err := requireID("iso2", iso2); err != nil {
		return nil, err
	}
	return callSlice[Route](ctx, r.http, request{method: http.MethodGet, path: "/v1/countries/" + escape(iso2) + "/routes"})
}

// Compliance returns a country's compliance rules
// (GET /v1/countries/{iso2}/compliance).
func (r *Countries) Compliance(ctx context.Context, iso2 string) (*CountryRules, error) {
	if err := requireID("iso2", iso2); err != nil {
		return nil, err
	}
	return call[CountryRules](ctx, r.http, request{method: http.MethodGet, path: "/v1/countries/" + escape(iso2) + "/compliance"})
}
